### PR TITLE
Vertex color layer and nodes

### PR DESCRIPTION
**Pull request by 3DSinghVFX**
1) This branch has Vertex Color Layer as mesh property like UVs. 
2) Two new nodes, **Get Mesh Vertex Color** and **Add Mesh Vertex Color**.

**Get Mesh Vertex Color**: This node takes mesh and the name of the vertex color layer as inputs. It gives the colors based on the color-mode (Loop, Vertex, Polygon).
Screenshot from 2019-12-18 22-56-35

**Add Mesh Vertex Color**: This node allows you to add a vertex color layer to the mesh. This node takes mesh, vertex color layer name, and color/colors (based on the color-mode e.g., Loop, Vertex, Polygon) as inputs. 
Screenshot from 2019-12-18 22-57-02

3) The Mesh Info, Mesh Object Input, and Mesh Object Output nodes are updated.

### CODE REVIEW

File: animation_nodes/nodes/mesh/add_mesh_vertex_color.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+from ... data_structures import Color, ColorList, VirtualColorList
+from .. color.c_utils import getLoopColorsFromVertexColors, getLoopColorsFromPolygonColors
+
+colorModeItems = [
+    ("LOOP", "Loop", "Get color of every loop vertex", "NONE", 0),
+    ("VERTEX", "Vertex", "Get color of every vertex", "NONE", 1),
+    ("POLYGON", "Polygon", "Get color of every polygon", "NONE", 2)    
+]
+
+class AddMeshVertexColorNode(bpy.types.Node, AnimationNode):
```
Comment (OmarEmaraDev) [inline animation_nodes/nodes/mesh/add_mesh_vertex_color.py]: Perhaps we should change the name to `InsertVertexColorLayerNode`, to be more uniform and to make it easier to search for. Do you agree?

File: animation_nodes/nodes/mesh/add_mesh_vertex_color.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            layout.prop(self, "colorMode", text = "")
+
+    def getExecutionFunctionName(self):
+        if self.useColorList:
+            return "execute_ColorsList"
+        else:
+            return "execute_SingleColor"
+
+    def execute_SingleColor(self, mesh, colorLayerName, color):
+        if mesh is None: return
```
Comment (OmarEmaraDev) [inline animation_nodes/nodes/mesh/add_mesh_vertex_color.py]: return None

File: animation_nodes/nodes/mesh/add_mesh_vertex_color.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+        defaultColor = Color((0, 0, 0, 1))
+        colorsList = VirtualColorList.create(color, defaultColor).materialize(len(mesh.polygons.indices))
+        
+        mesh.insertVertexColorLayer(colorLayerName, colorsList)
+        return mesh
+        
+    def execute_ColorsList(self, mesh, colorLayerName, colors):
+        if mesh is None: return None
+        if colorLayerName == "": return mesh
```
Comment (OmarEmaraDev) [inline animation_nodes/nodes/mesh/add_mesh_vertex_color.py]: We should probably inform the user using exception error handling.

File: animation_nodes/nodes/mesh/get_mesh_vertex_color.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        self.newInput("Mesh", "Mesh", "mesh")
+        self.newInput("Text", "Name", "colorLayerName")
+
+        self.newOutput("Color List", "Colors", "colors")        
+
+    def draw(self, layout):
+        layout.prop(self, "colorMode", text = "")
+
+    def execute(self, mesh, colorLayerName):
+        if mesh is None or colorLayerName not in mesh.getVertexColorLayerNames(): return ColorList()
```
Comment (OmarEmaraDev) [inline animation_nodes/nodes/mesh/get_mesh_vertex_color.py]: Inform the user using exception error handling here as well.

File: animation_nodes/nodes/mesh/get_mesh_vertex_color.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+        defaultColor = Color((0, 0, 0, 1))
+        
+        colorsList = ColorList(length = len(mesh.polygons.indices))
+        
+        for name, data in mesh.getVertexColorLayers():
+            if name == colorLayerName:
+                colorsList = data
+
```
Comment (OmarEmaraDev) [inline animation_nodes/nodes/mesh/get_mesh_vertex_color.py]: Add a method to the mesh that returns the vertex color layer for the given name.

File: animation_nodes/data_structures/meshes/mesh_data.pyx [hunk 6]
```diff
@@ -141,13 +154,18 @@ cdef class Mesh:
                 Vertices: {}
                 Edges: {}
```
Comment (OmarEmaraDev) [inline animation_nodes/data_structures/meshes/mesh_data.pyx]: Try replacing this with an f-string.

File: animation_nodes/nodes/mesh/add_mesh_vertex_color.py (hunk 1 above)
Comment (3DSinghVFX) [reply animation_nodes/nodes/mesh/add_mesh_vertex_color.py]: Omar, `Add Mesh Vertex Color` and  `Get Mesh Vertex Color` nodes do similar jobs, i.e., one node insert the vertex color layer and its colors, and the other gets the colors of a vertex color layer. I think if their names are similar then it would be nice. 

Add Mesh Vertex Color -> Insert Vertex Color Layer
Get Mesh Vertex Color -> Get Vertex Color Layer 

or

Add Mesh Vertex Color -> Mesh Vertex Color Output  or  Vertex Color Layer Output
Get Mesh Vertex Color -> Mesh Vertex Color Input  or  Vertex Color Layer Input

Comment (OmarEmaraDev) [reply animation_nodes/nodes/mesh/add_mesh_vertex_color.py]: Yes, sure. This looks good:

Add Mesh Vertex Color -> Insert Vertex Color Layer
Get Mesh Vertex Color -> Get Vertex Color Layer

Comment (3DSinghVFX) [reply animation_nodes/nodes/mesh/add_mesh_vertex_color.py]: Okay.

**Comment by 3DSinghVFX** [comment]
Omar, I'll remove the Vertex Color Input node. The reason I kept this node is because there is a bug in `color/c_utils.pyx`, so the Vertex Color Input node gives different colors,
RK1HfKVU jpg large


https://github.com/JacquesLucke/animation_nodes/blob/af8354fffcb9a05290fe81cebd3ad9ccbcf0c8d7/animation_nodes/nodes/color/c_utils.pyx#L64

When I change this to `addColor_Inplace(polygonColors.data + i, colors.get(index))` then Vertex Color Input node gives the same colors.

**Comment by OmarEmaraDev** [comment]
Sure, you may fix the bug in this pull request as well.

File: animation_nodes/data_structures/meshes/mesh_data.pyx (hunk 6 above)
Comment (3DSinghVFX) [reply animation_nodes/data_structures/meshes/mesh_data.pyx]: Omar, I have checked with the fresh AN compile build, there is no `runtime` warning. It is working fine.

Comment (OmarEmaraDev) [reply animation_nodes/data_structures/meshes/mesh_data.pyx]: Good. To clarify, I wanted you to using python's f-strings here. See https://www.python.org/dev/peps/pep-0498/

File: animation_nodes/nodes/mesh/add_mesh_vertex_color.py (hunk 2 above)
Comment (3DSinghVFX) [reply animation_nodes/nodes/mesh/add_mesh_vertex_color.py]: Added.

File: animation_nodes/nodes/mesh/add_mesh_vertex_color.py (hunk 3 above)
Comment (3DSinghVFX) [reply animation_nodes/nodes/mesh/add_mesh_vertex_color.py]: I have added the exception error handling.

File: animation_nodes/data_structures/meshes/mesh_data.pyx [hunk 7]
```diff
@@ -173,6 +191,13 @@ cdef class Mesh:
         for i in range(meshData.polygons.indices.length):
             self.polygons.indices.data[polygonIndicesOffset + i] += vertexOffset
 
+    def getVertexColorLayer(self, str colorLayerName):
+        cdef long i    
+        for i in range(len(self.getVertexColorLayers())):
+            if self.getVertexColorLayers()[i][0] == colorLayerName:
+                return self.getVertexColorLayers()[i]    
+            else:
+                return None
```
Comment (OmarEmaraDev) [inline animation_nodes/data_structures/meshes/mesh_data.pyx]: `vertexColorLayers` is an ordered dictionary, you can simply get the vertex color layer using subscript notation. Also, move this function below the `getVertexColorLayerNames`.

File: animation_nodes/nodes/mesh/get_mesh_vertex_color.py (hunk 4 above)
Comment (3DSinghVFX) [reply animation_nodes/nodes/mesh/get_mesh_vertex_color.py]: Here, it is also added the exception error handling.

File: animation_nodes/nodes/mesh/get_mesh_vertex_color.py (hunk 5 above)
Comment (3DSinghVFX) [reply animation_nodes/nodes/mesh/get_mesh_vertex_color.py]: Now this is a method which gives vertex color layer for the given name.

File: animation_nodes/data_structures/meshes/mesh_data.pyx (hunk 7 above)
Comment (3DSinghVFX) [reply animation_nodes/data_structures/meshes/mesh_data.pyx]: Is this correct way,
```
def getVertexColors(self, str colorLayerName):
        return self.vertexColorLayers.get(colorLayerName, None)
```
This function return the colors corresponds to the colorLayerName.

Comment (OmarEmaraDev) [reply animation_nodes/data_structures/meshes/mesh_data.pyx]: Yes, looks good.

**Comment by 3DSinghVFX** [comment]
Omar, I have used the f-strings, here is the code:
```
def __repr__(self):
        return textwrap.dedent(
        f"""AN Mesh Object:
        Vertices: {len(self.vertices)}
        Edges: {len(self.edges)}
        Polygons: {len(self.polygons)}
        UV Maps: {list(self.uvMaps.keys())}
        Vertex Color Layers: {list(self.vertexColorLayers.keys())}""")
```
I think we should change the `AN Mesh Object:` to `AN Mesh:`, is it not?.

**Comment by OmarEmaraDev** [comment]
I think *Object* here denotes an instance of the AN Mesh class, not a Blender object. So just leave it for now.

**Comment by 3DSinghVFX** [comment]
> I think _Object_ here denotes an instance of the AN Mesh class, not a Blender object. So just leave it for now.

Okay.

File: animation_nodes/data_structures/meshes/mesh_data.pyx (hunk 7 above)
Comment (3DSinghVFX) [reply animation_nodes/data_structures/meshes/mesh_data.pyx]: I have added this function.

**Comment by 3DSinghVFX** [comment]
Omar, In Insert Vertex Color Layer node, if we add a color layer of the same name which already there then does not notify the user. Should we add the exception for that,
screen
Or the viewer node is enough for user to see whether that color layer is already there or not.

**Comment by OmarEmaraDev** [comment]
I think we should do that. Raise an exception informing the user about the existing layer.

I will continue the review on Sunday as I don't have much time right now.

**Comment by 3DSinghVFX** [comment]
Okay.

No problem :)

**Comment by 3DSinghVFX** [comment]
Omar, I have noticed another problem with the order of color layers arises if we use Set Vertex Color with Color Layer Index option.

In this example, I have inserted two color layers. Then try setting the color to color layer _AN-Col2_ which has index zero with the Set Vertex Color node, 
Vertex_Color_Layer3

However, in this case, if include the existing color layer (_Col_) or insert a color layer before the _AN-Col2_ then the Index get change for _AN-Col2_ color layer for Set Vertex Color node,
Vertex_Color_Layer4

So, I think we should also use the name of the color layer in the Set Vertex Color node to resolve this problem. Do have any other idea?

File: animation_nodes/data_structures/meshes/mesh_data.pyx [hunk 8]
```diff
@@ -136,18 +152,21 @@ cdef class Mesh:
...
-                Polygons: {}
-                UV Maps: {}\
-            """.format(len(self.vertices), len(self.edges), len(self.polygons), list(self.uvMaps.keys())))
+        return textwrap.dedent(
+        f"""AN Mesh Object:
+        Vertices: {len(self.vertices)}
+        Edges: {len(self.edges)}
+        Polygons: {len(self.polygons)}
+        UV Maps: {list(self.uvMaps.keys())}
+        Vertex Color Layers: {list(self.vertexColorLayers.keys())}""")
```
Comment (OmarEmaraDev) [inline animation_nodes/data_structures/meshes/mesh_data.pyx]: Can't we use `getVertexColorLayerNames` here?

File: animation_nodes/nodes/mesh/get_vertex_color_layer.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+colorModeItems = [
+    ("LOOP", "Loop", "Get color of every loop vertex", "NONE", 0),
+    ("VERTEX", "Vertex", "Get color of every vertex", "NONE", 1),
+    ("POLYGON", "Polygon", "Get color of every polygon", "NONE", 2)    
+]
+
+class GetVertexColorLayerNode(bpy.types.Node, AnimationNode):
+    bl_idname = "an_GetVertexColorLayerNode"
+    bl_label = "Get Vertex Color Layer"
+    bl_width_default = 155
```
Comment (OmarEmaraDev) [inline animation_nodes/nodes/mesh/get_vertex_color_layer.py]: I would rather we don't use custom widths here.

File: animation_nodes/nodes/mesh/get_vertex_color_layer.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+        self.newInput("Mesh", "Mesh", "mesh", dataIsModified = True)
+        self.newInput("Text", "Name", "colorLayerName")
+
+        self.newOutput("Color List", "Colors", "colors")        
+
+    def draw(self, layout):
+        layout.prop(self, "colorMode", text = "")
+
+    def execute(self, mesh, colorLayerName):
+        if mesh is None: return ColorList()
```
Comment (OmarEmaraDev) [inline animation_nodes/nodes/mesh/get_vertex_color_layer.py]: We can use an exception here as well.

File: animation_nodes/nodes/mesh/get_vertex_color_layer.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+        if colorLayerName == "":
+            self.raiseErrorMessage("No Vertex Color Layer Name.")
```
Comment (OmarEmaraDev) [inline animation_nodes/nodes/mesh/get_vertex_color_layer.py]: -> Vertex color layer name can't be empty.

File: animation_nodes/nodes/mesh/get_vertex_color_layer.py [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+        
+        defaultColor = Color((0, 0, 0, 1))
+        colorsList = ColorList(length = len(mesh.polygons.indices))
+        
+        if mesh.getVertexColors(colorLayerName) is not None:
+            colorsList = mesh.getVertexColors(colorLayerName)
```
Comment (OmarEmaraDev) [inline animation_nodes/nodes/mesh/get_vertex_color_layer.py]: You are calling `mesh.getVertexColors` two times here.

File: animation_nodes/nodes/mesh/get_vertex_color_layer.py [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+        else:
+            self.raiseErrorMessage(f"No Vertex Color Layer of Name '{colorLayerName}'."
+                                   f" Layers: {', '.join(mesh.getVertexColorLayerNames())}")
```
Comment (OmarEmaraDev) [inline animation_nodes/nodes/mesh/get_vertex_color_layer.py]: This might be a little long for an error message. Also, `of name` -> `with name`.

File: animation_nodes/nodes/mesh/get_vertex_color_layer.py [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+            
+            colorsList = VirtualColorList.create(colorsList, defaultColor)
+            return getVertexColorsFromLoopColors(vertexCount, polygonIndices, colorsList)
+        elif self.colorMode == "POLYGON":
+            polygonIndices = mesh.polygons
+
+            colorsList = VirtualColorList.create(colorsList, defaultColor)
+            return getPolygonColorsFromLoopColors(polygonIndices, colorsList)
+        
+
```
Comment (OmarEmaraDev) [inline animation_nodes/nodes/mesh/get_vertex_color_layer.py]: New line at the end of file.

File: animation_nodes/nodes/mesh/get_vertex_color_layer.py (hunk 13 above)
Comment (3DSinghVFX) [reply animation_nodes/nodes/mesh/get_vertex_color_layer.py]: I added the `f" Layers: {', '.join(mesh.getVertexColorLayerNames())}"` to avoid guesswork or to check names with Viewer Node. Should I remove this line to make the error message short?

Comment (OmarEmaraDev) [reply animation_nodes/nodes/mesh/get_vertex_color_layer.py]: Yes. The Mesh repr method already provide those names.

File: animation_nodes/nodes/mesh/get_vertex_color_layer.py (hunk 10 above)
Comment (3DSinghVFX) [reply animation_nodes/nodes/mesh/get_vertex_color_layer.py]: Omar, For input mesh,
https://github.com/JacquesLucke/animation_nodes/blob/419e1e73431b17cb5cfa6cce76acc588321c5092/animation_nodes/nodes/mesh/get_vertex_color_layer.py#L24
due to this, I think mesh can't none, is it not?

File: animation_nodes/nodes/mesh/get_vertex_color_layer.py (hunk 11 above)
Comment (3DSinghVFX) [reply animation_nodes/nodes/mesh/get_vertex_color_layer.py]: It is replaced.

File: animation_nodes/nodes/mesh/get_vertex_color_layer.py (hunk 10 above)
Comment (OmarEmaraDev) [reply animation_nodes/nodes/mesh/get_vertex_color_layer.py]: An input mesh can be None. But if it is None, I would consider this an issue with the node the produced the None value. So maybe you are right, the None check is not needed here.

File: animation_nodes/nodes/mesh/get_vertex_color_layer.py (hunk 13 above)
Comment (3DSinghVFX) [reply animation_nodes/nodes/mesh/get_vertex_color_layer.py]: Now it looks like this `self.raiseErrorMessage(f"Mesh doesn't has '{colorLayerName}' Vertex Color Layer.")`. is it fine?

File: animation_nodes/nodes/mesh/get_vertex_color_layer.py (hunk 9 above)
Comment (3DSinghVFX) [reply animation_nodes/nodes/mesh/get_vertex_color_layer.py]: Removed.

File: animation_nodes/nodes/mesh/get_vertex_color_layer.py (hunk 13 above)
Comment (OmarEmaraDev) [reply animation_nodes/nodes/mesh/get_vertex_color_layer.py]: I think this is better:

> Mesh doesn't have a vertex color layer with the name '{colorLayerName}'.

File: animation_nodes/data_structures/meshes/mesh_data.pyx (hunk 8 above)
Comment (3DSinghVFX) [reply animation_nodes/data_structures/meshes/mesh_data.pyx]: Yes. I have changed it, `Vertex Colors: {self.getVertexColorLayerNames()}`.

File: animation_nodes/nodes/mesh/get_vertex_color_layer.py (hunk 13 above)
Comment (3DSinghVFX) [reply animation_nodes/nodes/mesh/get_vertex_color_layer.py]: Okay, here it is
Screenshot from 2019-12-23 00-24-14

File: animation_nodes/nodes/mesh/get_vertex_color_layer.py (hunk 12 above)
Comment (3DSinghVFX) [reply animation_nodes/nodes/mesh/get_vertex_color_layer.py]: It is changed.

File: animation_nodes/nodes/mesh/get_vertex_color_layer.py (hunk 14 above)
Comment (3DSinghVFX) [reply animation_nodes/nodes/mesh/get_vertex_color_layer.py]: Removed the extra line.

File: animation_nodes/nodes/mesh/get_vertex_color_layer.py (hunk 13 above)
Comment (OmarEmaraDev) [reply animation_nodes/nodes/mesh/get_vertex_color_layer.py]: Looks good.

**Comment by 3DSinghVFX** [comment]
Omar, How we are going to handle this problem: https://github.com/JacquesLucke/animation_nodes/pull/1256#issuecomment-568158138

**Comment by OmarEmaraDev** [comment]
Well, the user can always identify the vertex color by name by changing the mode from the advanced node settings. So I don't really consider this an options.

**Comment by 3DSinghVFX** [comment]
Okay.

File: animation_nodes/nodes/mesh/get_vertex_color_layer.py [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+            self.raiseErrorMessage("Vertex color layer name can't be empty.")
+        
+        defaultColor = Color((0, 0, 0, 1))
+        colorsList = ColorList(length = len(mesh.polygons.indices))
+        
+        getColorList = mesh.getVertexColors(colorLayerName)
+        if getColorList is not None:
+            colorsList = getColorList
+        else:
+            self.raiseErrorMessage(f"Mesh doesn't have a vertex color layer with the name '{colorLayerName}'.")
```
Comment (OmarEmaraDev) [inline animation_nodes/nodes/mesh/get_vertex_color_layer.py]:
```
colorsList = mesh.getVertexColors(colorLayerName)
if colorsList is None:
    raise
continue as usual without an else.
```

File: animation_nodes/nodes/mesh/get_vertex_color_layer.py [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+            vertexCount = len(mesh.vertices)
+            
+            colorsList = VirtualColorList.create(colorsList, defaultColor)
+            return getVertexColorsFromLoopColors(vertexCount, polygonIndices, colorsList)
+        elif self.colorMode == "POLYGON":
+            polygonIndices = mesh.polygons
+
+            colorsList = VirtualColorList.create(colorsList, defaultColor)
+            return getPolygonColorsFromLoopColors(polygonIndices, colorsList)
+
```
Comment (OmarEmaraDev) [inline animation_nodes/nodes/mesh/get_vertex_color_layer.py]: Setup your editor to add a new line at the end of file.

Comment (3DSinghVFX) [reply animation_nodes/nodes/mesh/get_vertex_color_layer.py]: Sorry. I'm using Visual Studio.

Comment (3DSinghVFX) [reply animation_nodes/nodes/mesh/get_vertex_color_layer.py]: I have setup a new line in my Visual Studio, this how it shows the code,
Screenshot from 2019-12-23 01-56-05

File: animation_nodes/nodes/mesh/get_vertex_color_layer.py (hunk 15 above)
Comment (3DSinghVFX) [reply animation_nodes/nodes/mesh/get_vertex_color_layer.py]: It is done.

File: animation_nodes/nodes/mesh/get_vertex_color_layer.py [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+    def draw(self, layout):
+        layout.prop(self, "colorMode", text = "")
+
+    def execute(self, mesh, colorLayerName):
+        if colorLayerName == "":
+            self.raiseErrorMessage("Vertex color layer name can't be empty.")
+        
+        defaultColor = Color((0, 0, 0, 1))
+        colorsList = ColorList(length = len(mesh.polygons.indices))
```
Comment (OmarEmaraDev) [inline animation_nodes/nodes/mesh/get_vertex_color_layer.py]: This is redundant.

File: animation_nodes/nodes/mesh/mesh_object_input.py [hunk 18]
```diff
@@ -6,6 +6,7 @@
 class MeshObjectInputNode(bpy.types.Node, AnimationNode):
     bl_idname = "an_MeshObjectInputNode"
     bl_label = "Mesh Object Input"
+    bl_width_default = 145
```
Comment (OmarEmaraDev) [inline animation_nodes/nodes/mesh/mesh_object_input.py]: I would prefer if we keep the default width here as well.

File: animation_nodes/nodes/mesh/mesh_object_input.py [hunk 19]
```diff
@@ -14,6 +15,7 @@ def create(self):
         self.newInput("Boolean", "Use World Space", "useWorldSpace")
         self.newInput("Boolean", "Use Modifiers", "useModifiers", value = False)
         self.newInput("Boolean", "Load UVs", "loadUVs", value = False)
+        self.newInput("Boolean", "Load Vertex Colors", "loadVertexColors", value = False)
```
Comment (OmarEmaraDev) [inline animation_nodes/nodes/mesh/mesh_object_input.py]: Can you hide both of those inputs by default?

Comment (3DSinghVFX) [reply animation_nodes/nodes/mesh/mesh_object_input.py]: Done.

File: animation_nodes/nodes/mesh/mesh_object_input.py (hunk 18 above)
Comment (3DSinghVFX) [reply animation_nodes/nodes/mesh/mesh_object_input.py]: Now it has the default width.

File: animation_nodes/nodes/mesh/get_vertex_color_layer.py (hunk 17 above)
Comment (3DSinghVFX) [reply animation_nodes/nodes/mesh/get_vertex_color_layer.py]: Removed.

File: animation_nodes/nodes/mesh/mesh_object_input.py [hunk 20]
```diff
@@ -33,7 +32,7 @@ def create(self):
         self.newOutput("Integer List", "Material Indices", "materialIndices")
 
         self.newOutput("Text", "Mesh Name", "meshName")
```
Comment (3DSinghVFX) [inline animation_nodes/nodes/mesh/mesh_object_input.py]: Omar, Mesh name output has a bug.

Comment (OmarEmaraDev) [reply animation_nodes/nodes/mesh/mesh_object_input.py]: What is the bug?

Comment (3DSinghVFX) [reply animation_nodes/nodes/mesh/mesh_object_input.py]: When I connect the viewer node to the mesh name output then it breaks the node-tree:
Screenshot from 2019-12-23 15-45-23

Comment (OmarEmaraDev) [reply animation_nodes/nodes/mesh/mesh_object_input.py]: Fixed in b442dfe.

Comment (3DSinghVFX) [reply animation_nodes/nodes/mesh/mesh_object_input.py]: Okay :)

**Comment by OmarEmaraDev** [comment]
Can you add the changes to the CHANGELOG file? Notice that we are still on the same sub version, so adjust the file appropriately. The node that was removed in this pull request should be removed from the CHANGELOG as well.

**Comment by 3DSinghVFX** [comment]
Okay, I'll update the CHANGELOG file.

**Comment by 3DSinghVFX** [comment]
I have added the changes about the insert and get vertex color layer nodes, and removed the `- Added Vertex Color Input node.`. Do I have to add anything else?

```
## 2.1.5

### Added

- Added *Include End Point* option to the *Float Range* node.
- Added collection option to *Execution Triggers*.
- Added *Sound Falloff* node.
- Added *Subprograms* to the search menu.
- Added missing nodes to the node menu
- Added *Collection Operations* node.
- Added *Rotations* output to the *Particles Data* node.
- Added Get Vertex Color Layer node.
- Added Insert Vertex Color Layer node.
```

**Comment by OmarEmaraDev** [comment]
No, look good.

File: CHANGELOG.md [hunk 21]
```diff
@@ -37,6 +36,8 @@
 - Added missing nodes to the node menu
 - Added *Collection Operations* node.
 - Added *Rotations* output to the *Particles Data* node.
+- Added Get Vertex Color Layer node.
+- Added Insert Vertex Color Layer node.
```
Comment (OmarEmaraDev) [inline CHANGELOG.md]: Why did you add them to 2.1.5? They should go to the Unreleased section.

Comment (3DSinghVFX) [reply CHANGELOG.md]: Sorry :) Now they are in the Unreleased section under the Added category.

**Comment by OmarEmaraDev** [comment]
Shouldn't the Insert Vertex Color Layer node have options for Loop, Vertex, and Polygon like the Set Vertex Color node?

**Comment by OmarEmaraDev** [comment]
My bad, just realized the options are there.

**Comment by 3DSinghVFX** [comment]
Thank you for helping me to understand the different aspects of AN :)
